Z/IO/BIO.hs

```diff
@@ -477,6 +484,26 @@ ioBIO f = \ k x ->
     case x of Just x' -> f x' >>= k . Just
               _ -> k EOF
 
+-- | BIO node from a pure filter.
+--
+-- BIO node made with this funtion are stateless, thus can be reused across chains.
+filter ::  (a -> Bool) -> BIO a a
+filter f k = go
+  where
+    go (Just a) = when (f a) $ k (Just a)
+    go Nothing = k Nothing
+
+-- | BIO node from an impure filter.
+--
+-- BIO node made with this funtion may not be stateless, it depends on if the IO function use
+filterM ::  (a -> IO Bool) -> BIO a a
```

Comment:
Is there some special reason to restrict the computation in the `IO` Monad? IMO it can be generalized to any Applicative for free according to now implementation. :thinking:

Reply:
I don't think that's possible since the `BIO` type restricted to `IO`.